File: java/5_LongestPalindromicSubstring.java
class Solution {
    public String longestPalindrome(String s) {
        if (s.length() == 0 || s == null) return s;
        int len = s.length();
        boolean[][] dp = new boolean[len + 1][len + 2];
        int res_i = 0, res_j = 0;

        for (int i = len - 2; i >= 0; i--) {
            dp[i][i] = true;
            for (int j = i + 1; j < len; j++) {
                dp[i][j] = s.charAt(i) == s.charAt(j) && (j - i < 3 || dp[i + 1][j - 1]);
                if (dp[i][j] && res_j - res_i < j - i) {
                    res_i = i;
                    res_j = j;
                }
            }
        }

        return s.substring(res_i, res_j + 1);
    }
}

File: c++/5_LongestPalindromicSubstring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        if (n == 0 || n == 1) return s;

        vector<vector<int>> dp(n, vector<int>(n));
        string res;
        
        for (int len = 0; len < n; len++) {
            for (int i = 0; i + len < n; i++) {
                int j = i + len;
                if (len == 0) {
                    dp[i][j] = 1;
                } else if (len == 1) {
                    dp[i][j] = (s[i] == s[j]);
                } else {
                    dp[i][j] = (s[i] == s[j]) && (dp[i+1][j-1]);
                }
                if (dp[i][j] && len + 1 > res.size()) {
                    res = s.substr(i, len+1);
                }
            }
        }
        
        return res;
    }
};
